### PR TITLE
Fix bug with basepath reservation

### PR DESCRIPTION
**Pull request by carlosmartinez**
Resolves [a bug](https://trello.com/c/CrtqBpmu/793-fix-bug-with-basepath-reservation-small#) and introduces some refactoring.

### CODE REVIEW

**Comment by kevindew** [comment]
Woo 👏  a PR on day one 🎉

File: app/models/path_reservation.rb [hunk 1]
```diff
@@ -1,18 +1,43 @@
 class PathReservation < ActiveRecord::Base
   validates :base_path, absolute_path: true
   validates :publishing_app, presence: true
-  validates_with PublishingAppValidator
 
-  def self.reserve_base_path!(base_path, publishing_app)
-    record = find_or_initialize_by(base_path: base_path)
-    ActiveRecord::Base.transaction do
-      record.update!(publishing_app: publishing_app)
+  class << self
```
Comment (danielroseman) [inline app/models/path_reservation.rb]: The styleguide recommends avoiding `class << self`, see third point under https://github.com/alphagov/styleguides/blob/master/ruby.md#classes

File: app/models/path_reservation.rb [hunk 2]
```diff
@@ -1,17 +1,43 @@
...
 
-  def self.reserve_base_path!(base_path, publishing_app)
-    record = find_or_initialize_by(base_path: base_path)
-    record.publishing_app = publishing_app
-    record.save!
-  rescue ActiveRecord::RecordNotUnique, PG::UniqueViolation
-    # If a path is already reserved by the same publishing app, ignore the error
-    record = find_by(base_path: base_path)
-    raise unless record.publishing_app == publishing_app
+  class << self
```
Comment (danielroseman) [inline app/models/path_reservation.rb]: The styleguide recommends avoiding `class << self`, see third point under https://github.com/alphagov/styleguides/blob/master/ruby.md#classes

**Comment by danielroseman** [comment]
One style comment, and the build has also failed on lint error. Otherwise :+1:

File: app/models/path_reservation.rb [hunk 3]
```diff
@@ -1,17 +1,43 @@
...
+        raise existing.already_reserved_error
+      end
+    end
 
-    record
+    private
+
+    def create_path_reservation(base_path, publishing_app)
+      ActiveRecord::Base.transaction do
+        create!(base_path: base_path, publishing_app: publishing_app)
```
Comment (dougdroper) [inline app/models/path_reservation.rb]: Is it necessary to wrap this in a transaction?

File: app/models/path_reservation.rb [hunk 4]
```diff
@@ -1,17 +1,41 @@
...
+    ActiveRecord::Base.transaction do
+      create!(base_path: base_path, publishing_app: publishing_app)
+    end
   rescue ActiveRecord::RecordNotUnique, PG::UniqueViolation
     # If a path is already reserved by the same publishing app, ignore the error
     record = find_by(base_path: base_path)
-    raise unless record.publishing_app == publishing_app
-
-    record
+    if record.already_associated_with?(publishing_app)
```
Comment (dougdroper) [inline app/models/path_reservation.rb]: is there a way to remove this duplication with `reserve_base_path!`

**Comment by cbaines** [comment]
For future reference, here is the [Git styleguide][1]. In particular, on this branch you don't always use the present tense, and don't always wrap to 72 characters.

[1]: https://github.com/alphagov/styleguides/blob/master/git.md

**Comment by cbaines** [comment]
(I've only just realised that the styleguide includes Git, otherwise I would have pointed this out at the time)

File: app/models/path_reservation.rb (hunk 2 above)
Comment (carlosmartinez) [inline app/models/path_reservation.rb]: good point. resolved.

File: app/models/path_reservation.rb (hunk 3 above)
Comment (carlosmartinez) [inline app/models/path_reservation.rb]: The transaction is kind of the substance of the bug fix!

File: app/models/path_reservation.rb (hunk 4 above)
Comment (carlosmartinez) [inline app/models/path_reservation.rb]: Good spot. Done.

**Pull request merged**